tox.ini

```diff
@@ -25,23 +25,33 @@ passenv =
 
 [testenv:docs]
 description = Build the Sphinx docs
-deps =
-    -r{toxinidir}/docs/requirements.txt
+deps = pip-tools
+commands_pre =
+    pip-compile --extra=docs -o docs/requirements.txt pyproject.toml
+    pip-sync {toxinidir}/docs/requirements.txt
```

Comment:
For some reason this `tox -e docs` fails for me, with this output (even after a `rm -rf .tox`):

```
$ tox -e docs
docs: commands_pre[0]> pip-compile --extra=docs -o docs/requirements.txt pyproject.toml
Backend subprocess exited when trying to invoke get_requires_for_build_wheel
Failed to parse /home/ben/w/operator/pyproject.toml
docs: exit 2 (3.59 seconds) /home/ben/w/operator> pip-compile --extra=docs -o docs/requirements.txt pyproject.toml pid=459533
  docs: FAIL code 2 (3.60=setup[0.02]+cmd[3.59] seconds)
  evaluation failed :( (3.63 seconds)
```

However, the first time I tried it it worked! (maybe due to existing stuff in `.tox/docs`?). 

Also, I noticed that when it did work, after running it I had local mods to `docs/requirements.txt` with it saying "This file was autogenerated ... with Python 3.11". Seems awkward to have the Python version in there.

Reply:
Hmm, if I blow away `.tox` and `docs/_build` then it works both first try and then second. If I blow them away, do a `tox -e docs` with `main` and then (without removing anything, other than what git does) switch to the branch and do `tox -e docs` again, it also works.

This might be because you're using (I think) Python 3.8 and I used 3.11. The readthedocs config is set to use Python 3.11, but unless we actually require that it's probably better to pin at whatever 3.8 supports.

I've regenerated it with Python 3.8. Does it work for you now?

> Seems awkward to have the Python version in there.

This is part of the `pip-compile` output - I don't see a config option to not have it there. We don't have to use `pip-tools` for this - I think it is currently the best tool that works well with both `pyproject.toml` and `tox`, but there are other options:

1. We could have it do the same thing as the readthedocs CI, `pip install .[docs]`. The only downside is that if you change a docs requirement in `pyproject.toml` and have an existing `.tox/docs` it won't pick up the change (this isn't an issue for the CI runner where it's always a fresh environment).
2. We could go back to having an `update_requirements.sh` script for the docs (but document that this time :smile:).
3. We could use something that's not `pip-tools` to generate the `requirements.txt` file (for our purposes it can be as simple as `python -c "import pytoml;print('\n'.join(pytoml.load(open('pyproject.toml'))['project']['optional-dependencies']['docs']))"`)
4. We could duplicate the dependency list in `tox.ini` and `pyproject.toml` and add some tooling to make sure they don't get out of sync.
5. Something else :)

Any preference?

Reply:
Weird, I couldn't figure out the exact reason for the failure out on my machine. I ended up cloning the project to a new, and then it worked, so I figured it was something cached in my other dir. I ended up doing a `git clean -d -x -f` and now it's working again...

Now that it's working, I think it's probably fine as is, thanks.